plugin/evm/block.go

```diff
@@ -80,75 +159,107 @@ func (b *Block) Height() uint64 {
 	return b.ethBlock.Number().Uint64()
 }
 
+var (
+	errRejectedParent = errors.New("rejected parent")
+)
+
 // Verify implements the snowman.Block interface
 func (b *Block) Verify() error {
+	vm := b.vm
+
 	// Only enforce a minimum fee when bootstrapping has finished
-	if b.vm.ctx.IsBootstrapped() {
+	if vm.ctx.IsBootstrapped() {
 		// Ensure the minimum gas price is paid for every transaction
 		for _, tx := range b.ethBlock.Transactions() {
 			if tx.GasPrice().Cmp(params.MinGasPrice) < 0 {
-				return errInvalidBlock
+				return errInvalidGas
 			}
 		}
 	}
 
-	vm := b.vm
+	ancestorIntf := b.Parent()
+	// Ensure that the parent was verified and inserted correctly.
+	ancestorID := ancestorIntf.ID()
+	ancestorHash := common.Hash(ancestorID)
+	if !vm.chain.BlockChain().HasBlock(ancestorHash, b.Height()-1) {
+		return errRejectedParent
+	}
+
+	// If the tx is an atomic tx, ensure that it doesn't conflict with any of
+	// its processing ancestry.
 	tx := vm.getAtomicTx(b.ethBlock)
 	if tx != nil {
-		pState, err := b.vm.chain.BlockState(b.Parent().(*Block).ethBlock)
+		// If the ancestor is unknown, then the parent failed verification when
+		// it was called.
+		// If the ancestor is rejected, then this block shouldn't be inserted
+		// into the canonical chain because the parent is will be missing.
+		if blkStatus := ancestorIntf.Status(); blkStatus == choices.Unknown || blkStatus == choices.Rejected {
+			return errRejectedParent
+		}
+		ancestor := ancestorIntf.(*Block)
```

Comment:
Can we do a type check here to avoid a potential panic?

```
ancestor, ok := ancestorIntf.(*Block)
if !ok {
    return fmt.Errorf("expected %s, parent of %s, to be *Block but is %T", ancestor.ID(), b.ID(), ancestorIntf)
}
```